### ISSUE TITLE
TF 2.1.0 crashes during shape inference

### ISSUE THREAD

**Issue by bioothod** [tensorflow/tensorflow#36542]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): standard docker image with ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 
- Python version: v2.1.0-rc2-17-ge5bf8de 2.1.0
- CUDA/cuDNN version: 10.1.243-1 / 7.6.4.38-1+cuda10.1
- GPU model and memory: titan rtx 24220MiB

**Describe the current behavior**
Using rather large complex model with lots of tf.TensorArray operations ends up with the crash. 100% reproducible, but overall codebase (python only) is quite large and input pipeline is tricky.

**Code to reproduce the issue**
This is the class I use to operate with tensor arrays, which seems to be a reason for crash.
It works in evaluation mode though
```
class RNNBatch:
    def __init__(self, rnn_layer, rnn_batch_size, true_words, true_lengths, training):
        self.rnn_batch_size = rnn_batch_size
        self.training = training
        self.dtype = tf.float32

        self.rnn_layer = rnn_layer
        max_sequence_len = rnn_layer.max_sequence_len
        dictionary_size = rnn_layer.dictionary_size

        self.true_words = true_words
        self.true_lengths = true_lengths

        self.written = 0
        self.rnn_processed_start = 0

        self.output_written = 0

    def run(self, arrays):
        selected_features = arrays['selected_features'].concat()

        batch_size = tf.shape(selected_features)[0]
        states_h = tf.zeros((batch_size, self.rnn_layer.num_rnn_units), dtype=self.dtype)
        states_c = tf.zeros((batch_size, self.rnn_layer.num_rnn_units), dtype=self.dtype)
        states = [states_h, states_c]

        tw = self.true_words[self.rnn_processed_start : self.rnn_processed_start + self.written, ...]
        tl = self.true_lengths[self.rnn_processed_start : self.rnn_processed_start + self.written, ...]

        out, out_ar = self.rnn_layer(selected_features, tw, tl, states, self.training)

        arrays['outputs'] = arrays['outputs'].write(self.output_written, out)
        arrays['outputs_ar'] = arrays['outputs'].write(self.output_written, out_ar)
        self.output_written += 1

        self.rnn_processed_start += self.written
        self.written = 0

        return arrays

    def feed_crop(self, cropped_features, arrays):
        arrays['selected_features'] = arrays['selected_features'].write(self.written, cropped_features)
        self.written += 1

        if self.written == self.rnn_batch_size:
            arrays = self.run(arrays)
            arrays['selected_features'] = tf.TensorArray(cropped_features.dtype, size=0, element_shape=tf.TensorShape([None] + [cropped_features.shape[1:]]))


        return arrays

    def return_values(self, arrays):
        if self.written != 0:
            arrays = self.run(arrays)

        return arrays['outputs'].concat(), arrays['outputs_ar'].concat()

```

And the use case is something like this:
```
images -> huge tensor of data ->
        rnn_batch = RNNBatch(self.rnn_layer, 64, true_words, true_lengths, training)
        arrays = {
            'selected_features': tf.TensorArray(dtype, size=0, dynamic_size=True, element_shape=tf.TensorShape([None, crop_size, crop_size, features_full.shape[3]])),
            'outputs': tf.TensorArray(dtype, size=0, dynamic_size=True, element_shape=tf.TensorShape([None, self.max_sequence_len, self.dictionary_size])),
            'outputs_ar': tf.TensorArray(dtype, size=0, dynamic_size=True, element_shape=tf.TensorShape([None, self.max_sequence_len, self.dictionary_size])),
        }

        for idx in tf.range(batch_size):
            for crop_idx in tf.range(num_crops):
                  arrays = rnn_batch.feed_crop(cropped_features, arrays)
```

**Other info / logs**
Full traceback: [tf.traceback.txt](https://github.com/tensorflow/tensorflow/files/4171246/tf.traceback.txt)
There is also a core file, gzipped size is about 749M.

What should be the next steps (besides rewriting above class) to debug this problem?

Attached stack trace which ends up like this:
```
Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
0x00007fff647051e0 in tensorflow::(anonymous namespace)::{lambda(tensorflow::shape_inference::InferenceContext*)#14}::_FUN(tensorflow::shape_inference::InferenceContext*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
(gdb) bt                                             
#0  0x00007fff647051e0 in tensorflow::(anonymous namespace)::{lambda(tensorflow::shape_inference::InferenceContext*)#14}::_FUN(tensorflow::shape_inference::InferenceContext*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#1  0x00007fff5d5a6104 in std::_Function_handler<tensorflow::Status (tensorflow::shape_inference::InferenceContext*), tensorflow::Status (*)(tensorflow::shape_inference::InferenceContext*)>::_M_invoke(std::_Any_d
ata const&, tensorflow::shape_inference::InferenceContext*&&) () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#2  0x00007fff5a018c22 in tensorflow::shape_inference::InferenceContext::Run(std::function<tensorflow::Status (tensorflow::shape_inference::InferenceContext*)> const&) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/../libtensorflow_framework.so.2
#3  0x00007fff64342a14 in tensorflow::ShapeRefiner::RunShapeFn(tensorflow::Node const*, tensorflow::OpRegistrationData const*, tensorflow::ExtendedInferenceContext*) ()
   from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#4  0x00007fff643443ee in tensorflow::ShapeRefiner::AddNode(tensorflow::Node const*) () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#5  0x00007fff5df740f1 in TF_FinishOperation () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
#6  0x00007fff5d529dd6 in _wrap_TF_FinishOperation () from /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/_pywrap_tensorflow_internal.so
```

Unfortunately there is no debug package I could use to make better trace

**Comment by bioothod** [comment]
Bug happens during augraph tracing time and is related to the typo in the second string:
arrays['outputs'] = arrays['outputs'].write(self.output_written, out)
arrays['outputs_ar'] = arrays['outputs'].write(self.output_written, out_ar)

**Comment by gadagashwini** [comment]
@bioothod, Can you provide the complete code to replicate the reported issue. Thanks!

**Comment by gadagashwini** [comment]
@bioothod, Any update

**Comment by bioothod** [comment]
This is a bit tricky process to run this large chunk of code, I can put both model and data online, but can you instead release package with debug symbols? It will be order of magnitude faster getting that I have all setup handy, I will also make a core

**Comment by ymodak** [comment]
Have you considered using [tensorboard debugger](https://github.com/tensorflow/tensorboard/tree/master/tensorboard/plugins/debugger#the-debugger-dashboard) for this case?

**Comment by bioothod** [comment]
Hmm, how can it help in catching segmentation fault?

Let me try to export the whole pipeline for this, it should not be hard to reproduce when everything is in place

**Comment by bioothod** [comment]
I will close it for now, because it looks like this should be fixed here: https://github.com/tensorflow/tensorflow/issues/33862
